beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/W3CPropagationCodec.java

```diff
@@ -0,0 +1,132 @@
+package io.honeycomb.beeline.tracing.propagation;
+
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.honeycomb.beeline.tracing.ids.W3CTraceIdProvider;
+
+import static io.honeycomb.libhoney.utils.ObjectUtils.isNullOrEmpty;
+
+/**
+ * Codec that can decode/encode trace context based on the W3C http header ('traceparent').
+ * Prefer the {@link Propagation#w3c()} factory method to retrieve an instance.
+ * <p>
+ * This implementation does not support the 'tracecontext' http header so no tracefields can be extracted or
+ * encoded.
+ * <p>
+ * The design of this class avoids throwing exceptions in favour of logging warnings and returning null on encode
+ * or an "empty context" on decode.
+ *
+ * <h1>Thread-safety</h1>
+ * Instances of this class are thread-safe and can be shared.
+ */
+public class W3CPropagationCodec implements PropagationCodec<String> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(W3CPropagationCodec.class);
+    private static final W3CPropagationCodec INSTANCE = new W3CPropagationCodec();
+
+    // @formatter:off
+    public static final String W3C_TRACEPARENT_HEADER       = "traceparent";
+    public static final String W3C_TRACESTATE_HEADER        = "tracestate";
```

Comment:
I noticed we're not doing anything with the tracestate header in this implementation. Is that by design? I'm okay not making use of the information, but in the example where a request comes in with both a traceparent and a tracestate header, and we want to instrument an outgoing call with W3C Trace Context headers, will we be able to maintain the tracestate value?

Reply:
Yeah, for now I've omitted parsing `tracecontext` as the current propagation codec design only allows reading / writing to a single HTTP header (`traceparent` in this case). When we move to a design where the all HTTP headers can be passed to the propagation codecs, we can extend the W3C codec to tease more information out.

I'll remove the `tracestate` header definition for now, and re-add when we come to that work :+1: